### PR TITLE
EFX tplg2

### PR DESCRIPTION
**Pull request by juimonen**

### CODE REVIEW

**Review by singalsu — APPROVED** [review]
Works well, thanks Jaska!

File: tools/topology/topology2/include/components/eqfir.conf [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		type "integer"
+	}
+
+	# Include common widget attributes definition
+	<include/components/widget-common.conf>
+
+	#
+	# Type of processing object
+	#
+	DefineAttribute."process_type" {
```
Comment (ranj063) [inline tools/topology/topology2/include/components/eqfir.conf:39]: why do we need this for IPC4?

File: tools/topology/topology2/include/components/eqfir/passthrough.conf [hunk 2]
```diff
@@ -0,0 +1,14 @@
+Object.Base.data."fir_eq" {
```
Comment (ranj063) [inline tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: why is this called passthrough.conf?

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		!constructor [
+			"index"
+		]
+
+		!immutable [
+			"direction"
+		]
+
+		#
+		# mixout-gain-dai-copier-playback objects instantiated within the same alsaconf node must have
```
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:44]: typo in name of the pipeline

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		# unique pipeline_id attribute
```
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:45]: this doesnt match the attribute name, its instance isnt it?

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+			Object.Base.audio_format.2 {
+				in_bit_depth		32
+				in_valid_bit_depth	32
+				out_bit_depth		32
+				out_valid_bit_depth	32
+			}
+
+			Object.Control.bytes."1" {
+				IncludeByKey.EFX_FIR_PARAMS {
```
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:113]: @juimonen @aiChaoSONG has a patch coming up for including small snippets of conf nodes without including a new file. We should wait until that is merged

File: tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	HDA.0 {
+		name $HDA_ANALOG_DAI_NAME
+		id 4
+		default_hw_conf_id 4
+		Object.Base.hw_config.HDA0 {}
+		direction duplex
+	}
+}
+Object.Pipeline {
+	mixout-gain-fir-dai-copier-playback.0 {
```
Comment (ranj063) [inline tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf:27]: The only difference between this one and the cavs-mixin-mixout-hda.conf is what pipeline is used in the backend. Can we make this conditional based on a variable definition?

File: tools/topology/topology2/include/components/eqfir/passthrough.conf (hunk 2 above)
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: so it is a flat response, does nothing - so "passthrough"

File: tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf (hunk 6 above)
Comment (juimonen) [reply tools/topology/topology2/cavs-mixin-mixout-fir-hda.conf:27]: there's also the byte control name definition... so diff is still small, but not only conf name. I can try to add some conditional logic.

File: tools/topology/topology2/include/components/eqfir.conf (hunk 1 above)
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf:39]: hmm ok, need to test without

File: tools/topology/topology2/include/components/eqfir/passthrough.conf (hunk 2 above)
Comment (singalsu) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf:1]: It's actually a blob that sets every channel filter index (preset) to -1. It instructs the EQ to pass that channel directly to output. When every channel is such the process() happens when mem copy operations for entire period.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf (hunk 4 above)
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf]: fixed

File: tools/topology/topology2/include/components/eqfir.conf (hunk 1 above)
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf]: removed, works without this.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf (hunk 5 above)
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf:113]: so we actually want to have the eq blobs in separate files as they can be produced with octave tools.

File: tools/topology/topology2/development/CMakeLists.txt [hunk 7]
```diff
@@ -23,6 +23,15 @@ PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-adl-nocodec.bin,DEEPBUFFER_FW_DMA_MS=1
 # SSP topology for MTL
 "cavs-nocodec\;sof-mtl-nocodec\;PLATFORM=mtl,NUM_DMICS=2,\
 PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-mtl-nocodec.bin,DEEPBUFFER_FW_DMA_MS=100"
+
+# CAVS HDA topology with mixer-based fir eq pipelines for HDA and passthrough pipelines for HDMI
+"sof-hda-generic\;sof-hda-fir-generic\;HDA_CONFIG=fir,USE_CHAIN_DMA=true,DEEPBUFFER_FW_DMA_MS=100,EFX_FIR_PARAMS=passthrough"
+"sof-hda-generic\;sof-hda-fir-generic-2ch\;\
```
Comment (ranj063) [inline tools/topology/topology2/development/CMakeLists.txt:29]: @juimonen @singalsu why is this a separate tplg? With IPC3 we had this in the regular HDA topologies isnt it?

Comment (juimonen) [reply tools/topology/topology2/development/CMakeLists.txt:29]: ok, I'll move to draft and move to generic

**Comment by juimonen** [comment]
moved to draft as I will make a change to include eq to hda-generic (it will be tested by CI). We have dependencies on kernel side
 https://github.com/thesofproject/linux/pull/4132/

**Comment by juimonen** [comment]
@ranj063 looking this again: I think PR as such is actually quite ok. You could change the "mainline" hda topology to use fir just by setting "HDA_CONFIG=fir" in cmakefile (currently it is HDA_CONFIG=mix). And I would like to have the possibility to move back and forth if needed. It is not very simple just rip off or add a component from topology (I've tried it couple of times and took me a while).

So just saying that I think this PR is good imho, we could change the mainline by follow up PR for cmakefile.

File: tools/topology/topology2/include/components/eqfir.conf [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+Class.Widget."eqfir" {
+	#
+	# Pipeline ID
+	#
+	DefineAttribute."index" {
+		type "integer"
+	}
+
+	#
+	# Unique instance for EQ IIR widget
```
Comment (lyakh) [inline tools/topology/topology2/include/components/eqfir.conf:27]: "FIR?"

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf (hunk 3 above)
Comment (juimonen) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-fir-dai-copier-playback.conf]: fixed

File: tools/topology/topology2/include/components/eqfir.conf (hunk 8 above)
Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir.conf]: fixed

**Comment by juimonen** [comment]
update: changed to generic EFX (end/post effect) pipeline, that has gain, iir and fir.

**Comment by singalsu** [comment]
Update with gain, IIR, FIR as EFX processing works here perfectly.  Patches needed in addition this are:

https://github.com/thesofproject/linux/pull/4132
https://github.com/thesofproject/sof/pull/6969
https://github.com/thesofproject/sof/pull/6992

Rimage is merged but need to make sure Zephyr build has https://github.com/thesofproject/rimage/pull/124

File: tools/topology/topology2/platform/intel/efx-default.conf [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Define {
+       EFX_FIR_PARAMS 'include/components/eqfir/passthrough.conf'
+       EFX_IIR_PARAMS 'include/components/eqiir/passthrough.conf'
+}
```
Comment (lgirdwood) [inline tools/topology/topology2/platform/intel/efx-default.conf:4]: missing newline at EOF.

Comment (juimonen) [reply tools/topology/topology2/platform/intel/efx-default.conf]: ok fixed

File: tools/topology/topology2/include/components/eqiir/passthrough.conf [hunk 10]
```diff
@@ -0,0 +1,18 @@
+Object.Base.data."iir_eq" {
+	bytes "
+		0x53,0x4f,0x46,0x34,0x58,0x00,0x00,0x00,
+		0x00,0x00,0x00,0x03,0x04,0xe0,0x00,0x03,
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply tools/topology/topology2/include/components/eqiir/passthrough.conf:4]: This should only come from the matlab tools, I really dont want to have to use ```bytes``` in topology2. i.e. we should use 
```
file "the_eq_coeff_file.bin"
```

[the remark this answers is not in the document]
Comment (juimonen) [reply tools/topology/topology2/include/components/eqiir/passthrough.conf]: now updated to reflect changes in:
https://github.com/thesofproject/linux/pull/4132

File: tools/topology/topology2/include/components/eqfir/passthrough.conf [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+	bytes "
+		0x53,0x4f,0x46,0x34,0x00,0x00,0x00,0x00,
+		0x38,0xe0,0x00,0x00,0x04,0x00,0x00,0x00,
```
Comment (ujfalusi) [inline tools/topology/topology2/include/components/eqfir/passthrough.conf:4]: 0x38, 0x00, 0x00, 0x00.
The 0x0e is wrong.

Comment (juimonen) [reply tools/topology/topology2/include/components/eqfir/passthrough.conf]: urgh indeed. fixed now.

**Comment by lgirdwood** [comment]
@juimonen looks like the dependencies are all merged. Still draft ?

**Review by lgirdwood — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**